File: src/curate/dedup.py
"""C6 — online HNSW near-duplicate detection. See TDD §4.5.

Index grows as the stream flows; each doc is compared against everything kept
before it. Parity (NFR5): fixed M/ef + single-threaded insertion order is
deterministic and matches the Rust impl.

Embed v4 returns L2-normalized vectors (verified OQ1), so cosine space needs no
extra normalization and the 0.92 threshold applies directly.
"""
from __future__ import annotations

import hnswlib
import numpy as np

from src.curate.minhash import MinHasher
from src.models import CurationConfig


class OnlineDedup:
    def __init__(self, config: CurationConfig, max_elements: int) -> None:
        self.config = config
        self.index = hnswlib.Index(space="cosine", dim=config.embed_dim)
        self.index.init_index(
            max_elements=max_elements,
            ef_construction=config.hnsw_ef_construction,
            M=config.hnsw_m,
        )
        self.index.set_ef(config.hnsw_ef_search)
        self.size = 0
        # Per-element MinHash signature, indexed by the internal label we assign
        # (== self.size at insert). None when a doc is added without a signature.
        self.signatures: list[np.ndarray | None] = []

    def check_and_add(
        self, doc_id: str, embedding: np.ndarray, signature: np.ndarray | None = None
    ) -> float | None:
        """Return similarity if near-duplicate (reject); None if kept + added.

        When `signature` is given, a cosine candidate is only rejected if the matched
        neighbour also clears the surface-Jaccard threshold — distinct-but-topically-
        similar docs (the false positives that emerge as the index grows) are kept.
        With signature=None this is the original cosine-only decision (back-compat).
        """
        if self.size > 0:
            labels, distances = self.index.knn_query(embedding, k=1)
            sim = 1.0 - float(distances[0][0])
            if sim >= self.config.dedup_cosine_threshold:
                neighbor_sig = self.signatures[int(labels[0][0])] if self.signatures else None
                if signature is None or neighbor_sig is None:
                    return sim  # cosine-only decision
                if MinHasher.jaccard(signature, neighbor_sig) >= self.config.dedup_surface_threshold:
                    return sim  # confirmed near-duplicate
                # else: embedding-similar but surface-distinct → false positive, keep
        self.index.add_items(embedding, self.size)
        self.signatures.append(signature)  # aligns with the internal label == self.size
        self.size += 1
        return None

File: rust/src/dedup.rs
//! Online HNSW dedup — mirror of src/curate/dedup.py. TDD §4.5, §16.
//!
//! Same M / ef / cosine space + single-threaded insertion order as hnswlib so
//! decisions are deterministic and match the Python reference (SC6). OQ4 resolved:
//! `hnsw_rs` (not `instant-distance`, which builds once and can't insert online
//! the way an incremental dedup stream needs).
//!
//! Embed v4 vectors are L2-normalized (OQ1), so cosine distance = 1 - dot and the
//! 0.90 threshold applies directly.

use hnsw_rs::prelude::*; // re-exports the anndists Distance trait + Hnsw

use crate::minhash::MinHasher;

const M: usize = 16;
const MAX_LAYER: usize = 16;
const EF_CONSTRUCTION: usize = 200;
const EF_SEARCH: usize = 64;

/// Cosine distance for already-L2-normalized vectors (OQ1): 1 - dot. Skips the
/// per-eval norm recomputation that hnsw_rs's built-in DistCosine pays, which
/// dominated the Rust hot path. sim = 1 - distance = dot, matching hnswlib's
/// cosine space on normalized input (parity, SC6).
#[derive(Clone)]
struct DistNormDot;

impl Distance<f32> for DistNormDot {
    fn eval(&self, a: &[f32], b: &[f32]) -> f32 {
        1.0 - a.iter().zip(b).map(|(x, y)| x * y).sum::<f32>()
    }
}

pub struct OnlineDedup<'a> {
    index: Hnsw<'a, f32, DistNormDot>,
    threshold: f32,
    surface_threshold: f32,
    signatures: Vec<Vec<u64>>, // per-element MinHash, indexed by the inserted data id
    size: usize,
}

impl<'a> OnlineDedup<'a> {
    pub fn new(threshold: f32, surface_threshold: f32, max_elements: usize) -> Self {
        let index = Hnsw::new(M, max_elements.max(1), MAX_LAYER, EF_CONSTRUCTION, DistNormDot {});
        Self { index, threshold, surface_threshold, signatures: Vec::new(), size: 0 }
    }

    /// Return Some(similarity) if near-duplicate (reject); None if kept + added.
    ///
    /// A cosine candidate is only rejected if the matched neighbour also clears the
    /// surface-Jaccard threshold — distinct-but-topically-similar docs are kept
    /// (mirror of src/curate/dedup.py; holds precision as the index grows).
    pub fn check_and_add(&mut self, emb: &[f32], sig: &[u64]) -> Option<f32> {
        if self.size > 0 {
            let neighbours = self.index.search(emb, 1, EF_SEARCH);
            if let Some(nb) = neighbours.first() {
                let sim = 1.0 - nb.distance;
                if sim >= self.threshold
                    && MinHasher::jaccard(sig, &self.signatures[nb.d_id]) >= self.surface_threshold
                {
                    return Some(sim);
                }
            }
        }
        self.index.insert((emb, self.size));
        self.signatures.push(sig.to_vec());
        self.size += 1;
        None
    }
}
